### ISSUE TITLE
sql: creating an index, user experienced OOM

### ISSUE THREAD

**Issue by roncrdb** [cockroachdb/cockroach#36381]
**Describe the problem**

User ran into OOM while creating an index, also experiencing some slight load from tests.

```
[10782031.209476] Out of memory: Kill process 25243 (cockroach) score 915 or sacrifice child
[10782031.210847] Killed process 25243 (cockroach) total-vm:118637456kB, anon-rss:62702344kB, file-rss:0kB, shmem-rss:0kB
```

Here are the top sources of Memory allocation from the node: 

```File: cockroach
Type: inuse_space
Time: Apr 1, 2019 at 5:41am (EDT)
Showing nodes accounting for 4534.31MB, 96.66% of 4691.17MB total
Dropped 406 nodes (cum <= 23.46MB)
      flat  flat%   sum%        cum   cum%
 4352.75MB 92.79% 92.79%  4352.75MB 92.79%  github.com/cockroachdb/cockroach/vendor/github.com/golang/leveldb/memfs.(*file).Write
      64MB  1.36% 94.15%       64MB  1.36%  github.com/cockroachdb/cockroach/vendor/github.com/andy-kimball/arenaskl.NewArena (inline)
   27.03MB  0.58% 94.73%    49.53MB  1.06%  github.com/cockroachdb/cockroach/pkg/storage.newReplica
   26.54MB  0.57% 95.29%    26.54MB  0.57%  github.com/cockroachdb/cockroach/pkg/storage/raftentry.realloc
      24MB  0.51% 95.80%    27.50MB  0.59%  github.com/cockroachdb/cockroach/pkg/sql/sqlbase.EncodeSecondaryIndex
   20.99MB  0.45% 96.25%    48.49MB  1.03%  github.com/cockroachdb/cockroach/pkg/sql/backfill.(*IndexBackfiller).BuildIndexEntriesChunk
      18MB  0.38% 96.64%    54.89MB  1.17%  github.com/cockroachdb/cockroach/vendor/go.etcd.io/etcd/raft.newRaft
    0.50MB 0.011% 96.65%    55.39MB  1.18%  github.com/cockroachdb/cockroach/vendor/go.etcd.io/etcd/raft.NewRawNode
    0.50MB 0.011% 96.66%    59.58MB  1.27%  github.com/cockroachdb/cockroach/pkg/storage.(*Replica).withRaftGroupLocked
```

[DEBUG ZIP](https://docs.google.com/document/d/1mWGjdHXT6a4ceH1yxTgeLYjICg-qViQcU0fT5wbT-2o/edit?usp=sharing)

**Expected behavior**
Index should complete without OOM

**Additional data / screenshots**
User has been trying to create an index, and has run into multiple issues, may be related to [#34878](https://github.com/cockroachdb/cockroach/issues/34878)

**Environment:**
 - CockroachDB v19.1.0-beta.20190318

**Comment by awoods187** [comment]
cc @vivekmenezes

**Comment by jordanlewis** [comment]
@roncrdb can you attach just the memory profile to this issue? Thanks!

**Comment by jordanlewis** [comment]
Yes, thanks!

Pasting the major allocation source along with its call stack for reference. @dt, this seems like the same thing you were looking at earlier, right?

![image](https://user-images.githubusercontent.com/43821/55355168-91884d00-5495-11e9-9da1-bf592df03925.png)

**Comment by roncrdb** [comment]
Looks like the link above to the debug zip isn't working, [here's a new link to the zip](https://drive.google.com/file/d/1dYecCGXaAKWyApu7rcwd0mdLTDd-Rj46/view?usp=sharing)

**Comment by jordanlewis** [comment]
Here's a stack trace from the unhappy node:

```
github.com/cockroachdb/cockroach/pkg/storage/bulk.addSplitSSTable(0x39d8080, 0xc02664e960, 0xc00064e580, 0xc1e0bb0000, 0xa5ee2a, 0xa5ee2a, 0xc1350368a0, 0x54, 0x60, 0xc1bf1bd3e0, ...)
    /go/src/github.com/cockroachdb/cockroach/pkg/storage/bulk/sst_batcher.go:285 +0x916
github.com/cockroachdb/cockroach/pkg/storage/bulk.AddSSTable(0x39d8080, 0xc02664e960, 0xc00064e580, 0xc1350368a0, 0x54, 0x60, 0xc135036960, 0x54, 0x54, 0xc1e0bb0000, ...)
    /go/src/github.com/cockroachdb/cockroach/pkg/storage/bulk/sst_batcher.go:210 +0x6a3
github.com/cockroachdb/cockroach/pkg/storage/bulk.addSplitSSTable(0x39d8080, 0xc02664e960, 0xc00064e580, 0xc1df6e8000, 0xa637eb, 0xa637eb, 0xc0fe357f80, 0x54, 0x60, 0xc1bf1bd320, ...)
    /go/src/github.com/cockroachdb/cockroach/pkg/storage/bulk/sst_batcher.go:285 +0x916
github.com/cockroachdb/cockroach/pkg/storage/bulk.AddSSTable(0x39d8080, 0xc02664e960, 0xc00064e580, 0xc0fe357f80, 0x54, 0x60, 0xc174ee8060, 0x54, 0x54, 0xc1df6e8000, ...)
    /go/src/github.com/cockroachdb/cockroach/pkg/storage/bulk/sst_batcher.go:210 +0x6a3
github.com/cockroachdb/cockroach/pkg/storage/bulk.addSplitSSTable(0x39d8080, 0xc02664e960, 0xc00064e580, 0xc1de214000, 0xa6897e, 0xa6897e, 0xc11278ec60, 0x53, 0x60, 0xc1bf1bd260, ...)
    /go/src/github.com/cockroachdb/cockroach/pkg/storage/bulk/sst_batcher.go:285 +0x916
github.com/cockroachdb/cockroach/pkg/storage/bulk.AddSSTable(0x39d8080, 0xc02664e960, 0xc00064e580, 0xc11278ec60, 0x53, 0x60, 0xc11278ed20, 0x54, 0x54, 0xc1de214000, ...)
    /go/src/github.com/cockroachdb/cockroach/pkg/storage/bulk/sst_batcher.go:210 +0x6a3
github.com/cockroachdb/cockroach/pkg/storage/bulk.addSplitSSTable(0x39d8080, 0xc02664e960, 0xc00064e580, 0xc1dcd38000, 0xa6d903, 0xa6d903, 0xc1c1e4e840, 0x53, 0x60, 0xc1bf1bd1a0, ...)
    /go/src/github.com/cockroachdb/cockroach/pkg/storage/bulk/sst_batcher.go:285 +0x916
github.com/cockroachdb/cockroach/pkg/storage/bulk.AddSSTable(0x39d8080, 0xc02664e960, 0xc00064e580, 0xc1c1e4e840, 0x53, 0x60, 0xc1c1e4e900, 0x54, 0x54, 0xc1dcd38000, ...)
    /go/src/github.com/cockroachdb/cockroach/pkg/storage/bulk/sst_batcher.go:210 +0x6a3
github.com/cockroachdb/cockroach/pkg/storage/bulk.addSplitSSTable(0x39d8080, 0xc02664e960, 0xc00064e580, 0xc1db850000, 0xa72d12, 0xa72d12, 0xc1cfbc3620, 0x53, 0x60, 0xc0dbc05040, ...)
    /go/src/github.com/cockroachdb/cockroach/pkg/storage/bulk/sst_batcher.go:285 +0x916
github.com/cockroachdb/cockroach/pkg/storage/bulk.AddSSTable(0x39d8080, 0xc02664e960, 0xc00064e580, 0xc1cfbc3620, 0x53, 0x60, 0xc1cfbc36e0, 0x54, 0x54, 0xc1db850000, ...)
    /go/src/github.com/cockroachdb/cockroach/pkg/storage/bulk/sst_batcher.go:210 +0x6a3
github.com/cockroachdb/cockroach/pkg/storage/bulk.addSplitSSTable(0x39d8080, 0xc02664e960, 0xc00064e580, 0xc1da360000, 0xa775dc, 0xa775dc, 0xc1cf9b3200, 0x52, 0x60, 0xc1bf1bd0e0, ...)
    /go/src/github.com/cockroachdb/cockroach/pkg/storage/bulk/sst_batcher.go:285 +0x916
github.com/cockroachdb/cockroach/pkg/storage/bulk.AddSSTable(0x39d8080, 0xc02664e960, 0xc00064e580, 0xc1cf9b3200, 0x52, 0x60, 0xc1cf9b32c0, 0x54, 0x54, 0xc1da360000, ...)
    /go/src/github.com/cockroachdb/cockroach/pkg/storage/bulk/sst_batcher.go:210 +0x6a3
github.com/cockroachdb/cockroach/pkg/storage/bulk.addSplitSSTable(0x39d8080, 0xc02664e960, 0xc00064e580, 0xc1d8e64000, 0xa7c3ea, 0xa7c3ea, 0xc1bef0a540, 0x51, 0x60, 0xc1bf1bd020, ...)
    /go/src/github.com/cockroachdb/cockroach/pkg/storage/bulk/sst_batcher.go:285 +0x916
github.com/cockroachdb/cockroach/pkg/storage/bulk.AddSSTable(0x39d8080, 0xc02664e960, 0xc00064e580, 0xc1bef0a540, 0x51, 0x60, 0xc1bef0a600, 0x54, 0x54, 0xc1d8e64000, ...)
    /go/src/github.com/cockroachdb/cockroach/pkg/storage/bulk/sst_batcher.go:210 +0x6a3
github.com/cockroachdb/cockroach/pkg/storage/bulk.addSplitSSTable(0x39d8080, 0xc02664e960, 0xc00064e580, 0xc1d7960000, 0xa80bce, 0xa80bce, 0xc0cea76240, 0x53, 0x60, 0xc0a6173b60, ...)
    /go/src/github.com/cockroachdb/cockroach/pkg/storage/bulk/sst_batcher.go:285 +0x916
github.com/cockroachdb/cockroach/pkg/storage/bulk.AddSSTable(0x39d8080, 0xc02664e960, 0xc00064e580, 0xc0cea76240, 0x53, 0x60, 0xc0cea76300, 0x54, 0x54, 0xc1d7960000, ...)
    /go/src/github.com/cockroachdb/cockroach/pkg/storage/bulk/sst_batcher.go:210 +0x6a3
github.com/cockroachdb/cockroach/pkg/storage/bulk.addSplitSSTable(0x39d8080, 0xc02664e960, 0xc00064e580, 0xc1d6454000, 0xa8592e, 0xa8592e, 0xc176a86900, 0x53, 0x60, 0xc0a6173a40, ...)
    /go/src/github.com/cockroachdb/cockroach/pkg/storage/bulk/sst_batcher.go:285 +0x916
github.com/cockroachdb/cockroach/pkg/storage/bulk.AddSSTable(0x39d8080, 0xc02664e960, 0xc00064e580, 0xc176a86900, 0x53, 0x60, 0xc1767e6b40, 0x54, 0x54, 0xc1d6454000, ...)
    /go/src/github.com/cockroachdb/cockroach/pkg/storage/bulk/sst_batcher.go:210 +0x6a3
github.com/cockroachdb/cockroach/pkg/storage/bulk.addSplitSSTable(0x39d8080, 0xc02664e960, 0xc00064e580, 0xc1d4f3c000, 0xa8af1a, 0xa8af1a, 0xc0e143c540, 0x52, 0x60, 0xc0a6173980, ...)
    /go/src/github.com/cockroachdb/cockroach/pkg/storage/bulk/sst_batcher.go:285 +0x916
github.com/cockroachdb/cockroach/pkg/storage/bulk.AddSSTable(0x39d8080, 0xc02664e960, 0xc00064e580, 0xc0e143c540, 0x52, 0x60, 0xc0e143c600, 0x54, 0x54, 0xc1d4f3c000, ...)
    /go/src/github.com/cockroachdb/cockroach/pkg/storage/bulk/sst_batcher.go:210 +0x6a3
created by github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*Flow).startInternal
    /go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/flow.go:563 +0x342
```

... except there were about 30 recursive calls.

**Comment by vivekmenezes** [comment]
looks like AddSSTable has some retry logic that should only be applied at the highest level of the recursion

**Comment by dt** [comment]
@vivekmenezes we can/should restructure the helpers here a bit, but with #36394 all the calls should be using the same `[]byte` which should help at least avoid the OOM.

**Comment by dt** [comment]
OOM fixed in #36394.
